syntax = "proto3";
package TikTok;

// Data structure from im/fetch/ response
message WebcastResponse {
  repeated Message messages = 1;
  string cursor = 2;
  int32 fetchInterval = 3;
  int64 serverTimestamp = 4;
  string internalExt = 5;
  int32 fetchType = 6; // ws (1) or polling (2)
  WebsocketParam wsParam = 7;
  int32 heartbeatDuration = 8;
  bool needAck = 9;
  string wsUrl = 10;
}

message Message {
  string type = 1;
  bytes binary = 2;
}

message WebsocketParam {
  string name = 1;
  string value = 2;
}

// Message types depending on Message.tyoe
message WebcastControlMessage {
  int32 action = 2;
}

// Statistics like viewer count
message WebcastRoomUserSeqMessage {
  int32 viewer_count = 3;
}

message WebcastChatMessage {
  User user = 2;
  string comment = 3;
}

message WebcastMemberMessage {
  User user = 2;
  WebcastMessageEvent event = 1;
  int32 action_id = 10;
}

message WebcastGiftMessage {
  User user = 7;

  // New Data Structure since 04/2022
  int32 id = 2;
  int32 repeatCount = 5;
  int32 repeatEnd = 9;
  WebcastGiftMessageGiftDetail info = 15;
  WebcastGiftMessageRecipient recipient = 23;

}

// Taken from https://github.com/Davincible/gotiktoklive/blob/da4630622bc586629a53faae64e8c53509af29de/proto/tiktok.proto#L57
message WebcastGiftMessageRecipient {
  uint64 timestamp = 6;
  uint64 user_id = 8;
}

message WebcastGiftMessageGiftDetail {
  WebcastGiftMessageGiftImage image = 1;
  string name = 16;
  string description = 2;
  int32 type = 11;
  int32 diamond_count = 12;
}

message WebcastGiftMessageGiftImage {
  string url = 1;
}


// Battle start
message WebcastLinkMicBattle {
  repeated WebcastLinkMicBattleItems battleUsers = 10;
}

message WebcastLinkMicBattleItems {
  WebcastLinkMicBattleGroup battleGroup = 2;
}

message WebcastLinkMicBattleGroup {
  LinkUser user = 1;
}


// Battle status
message WebcastLinkMicArmies {
  repeated WebcastLinkMicArmiesItems battleItems = 3;
  int32 battleStatus = 7;
}

message WebcastLinkMicArmiesItems {
  uint64 hostUserId = 1;
  repeated WebcastLinkMicArmiesGroup battleGroups = 2;
}

message WebcastLinkMicArmiesGroup {
  repeated User users = 1;
  int32 points = 2;
}


// Follow & share event
message WebcastSocialMessage {
  User user = 2;
  WebcastMessageEvent event = 1;
}

// Like event (is only sent from time to time, not with every like)
message WebcastLikeMessage {
  User user = 5;
  WebcastMessageEvent event = 1;
  int32 likes = 2;
  int32 total_likes = 3;
}

// New question event
message WebcastQuestionNewMessage {
  QuestionDetails questionDetails = 2;
}

message QuestionDetails {
  string question = 2;
  User user = 5;
}

message WebcastMessageEvent {
  WebcastMessageEventDetails details = 8;
}

// Contains UI information
message WebcastMessageEventDetails {
  string display_type = 1;
  string label = 2;
}

message User {
  uint64 user_id = 1;
  string nickname = 3;
  ProfilePicture avatar = 9;
  UserDetails details = 22;
  string unique_id = 38;
  repeated UserBadgesAttributes badges = 64;
}

message LinkUser {
  uint64 user_id = 1;
  string nickname = 2;
  ProfilePicture avatar = 3;
  string unique_id = 4;
}

message ProfilePicture {
  repeated string urls = 1;
}

message UserDetails {
  int32 follow_role = 3;
}

message UserBadgesAttributes {
  int32 badge_scene_type = 3;
  repeated UserTextBadge text_badges = 21;
  repeated UserImageBadge image_badges = 20;
}

message UserTextBadge {
  string type = 2;
  string name = 3;
}


// Websocket incoming message structure
message WebcastWebsocketMessage {
  uint64 id = 2;
  string type = 7;
  bytes binary = 8;
}

// Websocket acknowledgment message
message WebcastWebsocketAck {
  uint64 id = 2;
  string type = 7;
}

// Source: Co-opted https://github.com/zerodytrash/TikTok-Livestream-Chat-Connector/issues/19#issuecomment-1074150342
message WebcastLiveIntroMessage {
  uint64 room_id = 2;
  string message = 4;
  User streamer = 5;
}

message SystemMessage {
  string description = 2;
}

message WebcastInRoomBannerMessage {
  string data = 2;
}

message ExtraRankData {
  string colour = 1;
  uint64 id = 4;
}

message RankData {
  string rank = 11; // No, I have no idea why they made the rank a string
}


message WeeklyRanking {
  string type = 1;
  string label = 2;
  RankData data = 4;
  ExtraRankData extra = 3;
}

message RankContainer {
  WeeklyRanking rankings = 3;
}


message WebcastRankUpdateMessage {
  RankContainer data = 2;
}

message HourlyRankContainer {
  WeeklyRanking rankings = 4;
}

message WebcastHourlyRankMessage {
  HourlyRankContainer data = 2;
}

message UserImageBadge {
  int32 display_type = 1;
  UserImageBadgeImage image = 2;
}

message UserImageBadgeImage {
  string url = 1;
}

message WebcastEmoteChatMessage {
  User user = 2;
  EmoteDetails emote = 3;
}

message EmoteDetails {
  string emoteId = 1;
  EmoteImage image = 2;
}

message EmoteImage {
  string url = 1;
}

// Envelope (treasure boxes)
// Taken from https://github.com/ThanoFish/TikTok-Live-Connector/blob/9b215b96792adfddfb638344b152fa9efa581b4c/src/proto/tiktokSchema.proto
message WebcastEnvelopeMessage {
  TreasureBoxData treasureBoxData = 2;
  TreasureBoxUser treasureBoxUser = 1;
}

message TreasureBoxUser {
  TreasureBoxUser2 user2 = 8;
}

message TreasureBoxUser2 {
  repeated TreasureBoxUser3 user3 = 4;
}

message TreasureBoxUser3 {
  TreasureBoxUser4 user4 = 21;
}

message TreasureBoxUser4 {
  User user = 1;
}

message TreasureBoxData {
  uint32 coins = 5;
  uint32 openable = 6;
  uint64 timestamp = 7;
}

message RoomMessage {
  string description = 2;
}